tests/e2e/e2e_setup_test.go

```diff
@@ -391,84 +391,116 @@ func noRestart(config *docker.HostConfig) {
-	s.voteProposal(s.chains[0])
-	s.voteProposal(s.chains[1])
-
+	var wg sync.WaitGroup
+	wg.Add(2)
+	// use goroutines to vote on both chains concurrently (no possibility of sequence mismatch)
+	go s.voteProposal(s.chains[0], &wg)
+	go s.voteProposal(s.chains[1], &wg)
+	wg.Wait()
 	// wait till all chains halt at upgrade height
-	for _, chain := range s.chains {
-		for i := range chain.Validators {
-			s.T().Logf("waiting to reach upgrade height on %s validator container: %s", chain.ChainMeta.Id, s.valResources[chain.ChainMeta.Id][i].Container.ID)
+	for _, c := range s.chains {
+		if c.ChainMeta.Id == chain.ChainAID {
+			propHeight = propHeightA
+		} else {
+			propHeight = propHeightB
+		}
+		for i := range c.Validators {
+			// use counter to ensure no new blocks are being created
+			counter := 0
+			s.T().Logf("waiting to reach upgrade height on %s validator container: %s", s.valResources[c.ChainMeta.Id][i].Container.Name[1:], s.valResources[c.ChainMeta.Id][i].Container.ID)
 			s.Require().Eventually(
 				func() bool {
-					out := s.chainStatus(s.valResources[chain.ChainMeta.Id][i].Container.ID)
-					var syncInfo syncInfo
-					json.Unmarshal(out, &syncInfo)
-					if syncInfo.SyncInfo.LatestHeight != "75" {
-						s.T().Logf("current block height is %v, waiting for block 75 container: %s", syncInfo.SyncInfo.LatestHeight, s.valResources[chain.ChainMeta.Id][i].Container.ID)
+					currentHeight := s.chainHeight(s.valResources[c.ChainMeta.Id][i].Container.ID)
+					if currentHeight != propHeight {
+						s.T().Logf("current block height on %s is %v, waiting for block %v container: %s", s.valResources[c.ChainMeta.Id][i].Container.Name[1:], currentHeight, propHeight, s.valResources[c.ChainMeta.Id][i].Container.ID)
 					}
-					return syncInfo.SyncInfo.LatestHeight == "75"
+					if currentHeight > propHeight {
+						panic("chain did not halt at upgrade height")
+					}
+					if currentHeight == propHeight {
+						counter++
+					}
+					return counter == 3
 				},
-				2*time.Minute,
-				5*time.Second,
+				5*time.Minute,
+				time.Second,
 			)
-			s.T().Logf("reached upgrade height on %s validator container: %s", chain.ChainMeta.Id, s.valResources[chain.ChainMeta.Id][i].Container.ID)
+			s.T().Logf("reached upgrade height on %s container: %s", s.valResources[c.ChainMeta.Id][i].Container.Name[1:], s.valResources[c.ChainMeta.Id][i].Container.ID)
 		}
 	}
 
 	// remove all containers so we can upgrade them to the new version
 	for _, chain := range s.chains {
 		for i := range chain.Validators {
-			s.Require().NoError(s.dkrPool.RemoveContainerByName(s.valResources[chain.ChainMeta.Id][i].Container.Name))
+			var opts docker.RemoveContainerOptions
+			opts.ID = s.valResources[chain.ChainMeta.Id][i].Container.ID
+			opts.Force = true
+			s.dkrPool.Client.RemoveContainer(opts)
+			s.T().Logf("removed container: %s", s.valResources[chain.ChainMeta.Id][i].Container.Name[1:])
 		}
 	}
+	// upgrade to locally compiled binary concurently
+	wg.Add(2)
```

Comment:
Why do we have to do this concurrently?

Also, we should avoid hardcoding and duplicating calls

2 == len(s.chains)

and `go s.upgradeContainers(s.chains[0], &wg)` can be done in a loop

Reply:
Same as before, if we have to upgrade 200 containers I would rather them be done at the same time instead of waiting for each to be finished since they are not dependent on each other.

Reply:
Similar idea to: https://github.com/osmosis-labs/osmosis/pull/1458/files/dace939d3b2461b62fbd5c407393b6905339a21d#r869717359